src/Shared/Server/infra/server_test.go

```diff
@@ -0,0 +1,206 @@
+	serverAddr = "127.0.0.1:4000"
+
+	server := New[any, any, any]()
+	server.HandleSPA()
+	go server.ListenAndServe()
+
+	client := http.Client{}
+
+	type testCase struct {
+		path   string
+		status int
+		body   []byte
+	}
+
+	tests := map[string]testCase{
+		"index.html": {
+			path:   "/",
+			status: http.StatusOK,
+			body:   readFile(path.Join("test", "resources", "index.html")),
+		},
+		"index.js": {
+			path:   "/index.js",
+			status: http.StatusOK,
+			body:   readFile(path.Join("test", "resources", "index.js")),
+		},
+		"invalid": {
+			path:   "/foo",
+			status: http.StatusNotFound,
+			body:   []byte("404 page not found\n"),
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			resp, err := client.Get("http://" + serverAddr + test.path)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if resp.StatusCode != test.status {
+				t.Fatalf("expected status %d, got %d", test.status, resp.StatusCode)
+			}
+
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer resp.Body.Close()
+
+			if !reflect.DeepEqual(body, test.body) {
+				t.Fatalf("expected %s, got %s", test.body, body)
+			}
+		})
+	}
+}
+
+func TestLog(t *testing.T) {
+	defaultStaticPath = path.Join("test", "resources")
+	serverAddr = "127.0.0.1:4001"
+
+	logChan := make(chan bool)
+	server := New[any, any, any]()
+	server.HandleLog("/backend/log", logChan)
+	go server.ListenAndServe()
+
+	client := http.Client{}
+
+	resp := put(client, "/backend/log", "enable")
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("expected status %d, got %d", http.StatusOK, resp.StatusCode)
+	}
+
+	select {
+	case enable := <-logChan:
+		if enable != true {
+			log.Fatalln("expected enable")
+		}
+	case <-time.After(time.Millisecond):
+		log.Fatalln("expected response")
+	}
+
+	resp = put(client, "/backend/log", "disable")
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("expected status %d, got %d", http.StatusOK, resp.StatusCode)
+	}
+
+	select {
+	case enable := <-logChan:
+		if enable != false {
+			log.Fatalln("expected disable")
+		}
+	case <-time.After(time.Millisecond):
+		log.Fatalln("expected response")
+	}
+
+	resp = put(client, "/backend/log", "foo")
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("expected status %d, got %d", http.StatusOK, resp.StatusCode)
+	}
+
+	select {
+	case <-logChan:
+		log.Fatalln("expected nothing")
+	case <-time.After(time.Millisecond):
+	}
+}
+
+func TestWebSocket(t *testing.T) {
+	defaultStaticPath = path.Join("test", "resources")
+	serverAddr = "127.0.0.1:4002"
+
+	server := New[any, any, any]()
+	server.HandleWebSocketData("/backend/data", func(interfaces.WebSocket, <-chan any) {})
+	server.HandleWebSocketMessage("/backend/message", func(interfaces.WebSocket, <-chan any) {})
+	server.HandleWebSocketOrder("/backend/order", func(interfaces.WebSocket, chan<- any) {})
+	go server.ListenAndServe()
+
+	_, resp := wsClient("ws://" + serverAddr + "/backend/data")
```

Comment:
put "make ws request and check status switching protocol" in a separate function